### PR TITLE
Magpietts small and Attention changes

### PR DESCRIPTION
**Pull request by subhankar-ghosh**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

1) Adds Small MagpieTTS yamls
2) multiplicative attention
3) hard prior for inference.

**Collection**: [Note which collection this PR will affect]

TTS

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: scripts/magpietts/infer_and_evaluate.py [hunk 1]
```diff
@@ -218,6 +221,7 @@ def run_inference(
         checkpoint_name = checkpoint_file.split("/")[-1].split(".ckpt")[0]
     elif nemo_file is not None:
         model_cfg = MagpieTTSModel.restore_from(nemo_file, return_config=True)
+        print(model_cfg)
```
Comment (blisc) [inline scripts/magpietts/infer_and_evaluate.py]: Do we need this line?

File: scripts/magpietts/infer_and_evaluate.py [hunk 2]
```diff
@@ -240,7 +244,7 @@ def run_inference(
     else:
         exp_name = ""
 
-    checkpoint_name = checkpoint_file.split("/")[-1].split(".ckpt")[0]
+    checkpoint_name = checkpoint_file.split("/")[-1].split(".ckpt")[0] if checkpoint_file is not None else checkpoint_name
```
Comment (blisc) [inline scripts/magpietts/infer_and_evaluate.py]: Can you remove this line? checkpoint_name is defined in the if else above, I don't know why this returned.

File: scripts/magpietts/infer_and_evaluate.py [hunk 3]
```diff
@@ -481,31 +485,28 @@ def run_inference(
...
-    parser.add_argument('--codecmodel_path', type=str, default="/datap/misc/checkpoints/12.5_FPS_causal_13codebooks_codecmodel.nemo", help="Path to codec model (used for FCD computation unless --disable_fcd is specified)")
-    parser.add_argument('--datasets', type=str, default="libri_unseen_test_12.5")
-    parser.add_argument('--base_exp_dir', type=str, default="/datap/misc/eosmountedresson/")
-    parser.add_argument('--draco_exp_dir', type=str, default="/lustre/fsw/llmservice_nemo_speechlm/users/pneekhara/gitrepos/experiments/NewT5TTS_FixedPosEmb/AllKernselSize3/EdressonCodecExperiments/")
-    parser.add_argument('--server_address', type=str, default="pneekhara@login-eos02.eos.clusters.nvidia.com")
-    parser.add_argument('--exp_names', type=str, default="koel_12.5_FPS_causal_13codebooks_codecmodel_context5sec_LTN1,koel_12.5_FPS_causal_13codebooks_codecmodel_context5sec_LTN3")
-    parser.add_argument('--local_ckpt_dir', type=str, default="/datap/misc/experiment_checkpoints/localtransformer")
+    parser.add_argument('--codecmodel_path', type=str, default=None, help="Path to codec model (used for FCD computation unless --disable_fcd is specified)")
```
Comment (blisc) [inline scripts/magpietts/infer_and_evaluate.py]:
```suggestion
    parser.add_argument('--codecmodel_path', type=str, default=None, help="Path to codec model")
```

File: scripts/magpietts/infer_and_evaluate.py [hunk 4]
```diff
@@ -582,55 +586,11 @@ def main():
                 checkpoint_file=None,
                 nemo_file=nemo_file,
             )
-    # Mode 3: Discover and run experiments from a base directory
```
Comment (blisc) [inline scripts/magpietts/infer_and_evaluate.py:585]: Did we confirm that this is now unused?

File: scripts/magpietts/README.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+HPARAM=<Path to .yaml file for MagpieTTS that you want to evaluate>
+CODEC=<Path to .nemo audio codec file for MagpieTTS >
+OUT_DIR=<Path to output directory for evaluation >
+
+python scripts/magpietts/infer_and_evaluate.py \
+--checkpoint_files ${CKPT} \
+--hparams_files ${HPARAM} \
+--codecmodel_path ${CODEC} \
+--out_dir ${OUT_DIR} \
+--batch_size 32 \
```
Comment (blisc) [inline scripts/magpietts/README.md]: Should we default batch size to 32?

File: nemo/collections/tts/modules/transformer_2501.py [hunk 6]
```diff
@@ -230,16 +230,11 @@ def attn_naive(
         # attn_prior or square mask or vanilla attention
         if attn_prior is not None:
```
Comment (blisc) [inline nemo/collections/tts/modules/transformer_2501.py:231]: We need a toggle to enable this at inference, but not during training

File: nemo/collections/tts/models/magpietts.py [hunk 7]
```diff
@@ -1483,25 +1484,28 @@ def construct_inference_prior(self, prior_epsilon, cross_attention_scores,
...
+                    _attn_prior[bidx, 0, min(text_time_step_attended[bidx]+4, _text_len - 1) ] = 1.0
+                    _attn_prior[bidx, 0, min(text_time_step_attended[bidx]+5, _text_len - 1) ] = 1.0
 
                 # Penalize timesteps that have been attended to more than 10 times
                 for _timestep in attended_timestep_counter[bidx]:
                     if attended_timestep_counter[bidx][_timestep] >= 10:
                         # This means the timestep has been attended to more than 10 times (To avoid getting stuck)
-                        _attn_prior[bidx, 0, _timestep] = prior_epsilon
+                        _attn_prior[bidx, 0, :_timestep+1] = prior_epsilon
+                        _attn_prior[bidx, 0, :_timestep+1] = prior_epsilon
```
Comment (blisc) [inline nemo/collections/tts/models/magpietts.py]:
```suggestion
                        _attn_prior[bidx, 0, :_timestep+1] = prior_epsilon
```

File: nemo/collections/tts/models/magpietts.py [hunk 8]
```diff
@@ -1483,25 +1484,28 @@ def construct_inference_prior(self, prior_epsilon, cross_attention_scores,
...
-                    _attn_prior[bidx, 0, text_time_step_attended[bidx]] = 0.8 # Slightly bias to continue moving forward. Not very important.
+                    _attn_prior[bidx, 0, max(1, text_time_step_attended[bidx]-1)] = 1.0 # Slight exposure to history for better pronounciation. Not very important.
+                    _attn_prior[bidx, 0, text_time_step_attended[bidx]] = 1.0 # Slightly bias to continue moving forward. Not very important.
                     _attn_prior[bidx, 0, min(text_time_step_attended[bidx]+1, _text_len - 1) ] = 1.0
-                    _attn_prior[bidx, 0, min(text_time_step_attended[bidx]+2, _text_len - 1) ] = 0.8
+                    _attn_prior[bidx, 0, min(text_time_step_attended[bidx]+2, _text_len - 1) ] = 1.0
+                    _attn_prior[bidx, 0, min(text_time_step_attended[bidx]+3, _text_len - 1) ] = 1.0
+                    _attn_prior[bidx, 0, min(text_time_step_attended[bidx]+4, _text_len - 1) ] = 1.0
+                    _attn_prior[bidx, 0, min(text_time_step_attended[bidx]+5, _text_len - 1) ] = 1.0
```
Comment (blisc) [inline nemo/collections/tts/models/magpietts.py]: Can we update this to match `attention_prior_lookahead_window` from infer and evaluate?

File: nemo/collections/tts/models/magpietts.py [hunk 9]
```diff
@@ -80,6 +80,7 @@ class MagpieTTSModel(ModelPT):
     """
 
     def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
+        print(f"cfg: {cfg}")
```
Comment (blisc) [inline nemo/collections/tts/models/magpietts.py]: Do we need this print statement?

File: examples/tts/conf/magpietts/magpietts_dc_en_tiny.yaml [hunk 10]
```diff
@@ -0,0 +1,175 @@
+name: Magpie-TTS-EN
```
Comment (blisc) [inline examples/tts/conf/magpietts/magpietts_dc_en_tiny.yaml]: When do we use this yaml?

File: scripts/magpietts/infer_and_evaluate.py (hunk 1 above)
Comment (subhankar-ghosh) [reply scripts/magpietts/infer_and_evaluate.py]: Removed

File: scripts/magpietts/infer_and_evaluate.py (hunk 2 above)
Comment (subhankar-ghosh) [reply scripts/magpietts/infer_and_evaluate.py]: Removed.

File: scripts/magpietts/infer_and_evaluate.py (hunk 3 above)
Comment (subhankar-ghosh) [reply scripts/magpietts/infer_and_evaluate.py]: Done

File: scripts/magpietts/infer_and_evaluate.py (hunk 4 above)
Comment (subhankar-ghosh) [reply scripts/magpietts/infer_and_evaluate.py:585]: Yes checked with Paarth

File: scripts/magpietts/README.md (hunk 5 above)
Comment (subhankar-ghosh) [reply scripts/magpietts/README.md]: Done

File: nemo/collections/tts/models/magpietts.py (hunk 7 above)
Comment (subhankar-ghosh) [reply nemo/collections/tts/models/magpietts.py]: Done

File: nemo/collections/tts/models/magpietts.py (hunk 8 above)
Comment (subhankar-ghosh) [reply nemo/collections/tts/models/magpietts.py]: Done

File: nemo/collections/tts/models/magpietts.py (hunk 9 above)
Comment (subhankar-ghosh) [reply nemo/collections/tts/models/magpietts.py]: Removed.

File: examples/tts/conf/magpietts/magpietts_dc_en_tiny.yaml (hunk 10 above)
Comment (subhankar-ghosh) [reply examples/tts/conf/magpietts/magpietts_dc_en_tiny.yaml]: This is the yaml for small version of magpietts

Comment (blisc) [reply examples/tts/conf/magpietts/magpietts_dc_en_tiny.yaml]: Why do we need this particular yaml compared to the lhotse one?

File: nemo/collections/tts/models/magpietts.py [hunk 11]
```diff
@@ -1483,25 +1483,24 @@ def construct_inference_prior(self, prior_epsilon, cross_attention_scores,
                     # Very short sentences, No Prior
                     _attn_prior[bidx, 0, :] = 1.0
```
Comment (blisc) [inline nemo/collections/tts/models/magpietts.py:1484]: What does `start_prior_after_n_audio_steps` do if we hard code this into our prior calculations? Do we need to remove this?

Comment (shehzeen) [reply nemo/collections/tts/models/magpietts.py:1484]: It is being used here: https://github.com/NVIDIA/NeMo/blob/69a7d544e4f08fa53074f7368c08db2bbfc2e14a/nemo/collections/tts/models/magpietts.py#L1680C17-L1680C85

I think it means to not apply the inference prior algorithm until `start_prior_after_n_audio_steps` decoding timesteps even for long sentences.

File: nemo/collections/tts/modules/transformer_2501.py [hunk 12]
```diff
@@ -167,6 +168,7 @@ def __init__(
         self.o_net = torch.nn.Linear(n_heads * self.d_head, d_model, bias=False)
         self.dropout = torch.nn.Dropout(p_dropout)
         self.use_cache = False
+        self.is_training = is_training
```
Comment (blisc) [inline nemo/collections/tts/modules/transformer_2501.py]: Use one of pytorch's built-in flags instead of adding a self arg

**Review by blisc — APPROVED** [review]
